core/src/main/java/org/elasticsearch/cluster/routing/OperationRouting.java

```diff
@@ -223,34 +219,29 @@ public ShardId shardId(ClusterState clusterState, String index, String id, @Null
     }
 
     static int generateShardId(IndexMetaData indexMetaData, @Nullable String id, @Nullable String routing) {
-        if (indexMetaData.isRoutingPartitionedIndex()) {
-            if (routing == null) {
-                throw new IllegalArgumentException("expected a routing value on partitioned index [" + indexMetaData.getIndex().getName() + "]");
-            }
+        final String effectiveRouting;
+        final int partitionOffset;
 
-            return generatePartitionedShardId(indexMetaData, routing, calculatePartitionOffset(indexMetaData, id));
+        if (routing == null) {
+            assert(!indexMetaData.isRoutingPartitionedIndex());
+            effectiveRouting = id;
         } else {
-            final int hash;
-
-            if (routing == null) {
-                hash = Murmur3HashFunction.hash(id);
-            } else {
-                hash = Murmur3HashFunction.hash(routing);
-            }
+            effectiveRouting = routing;
+        }
 
-            return calculateScaledShardId(indexMetaData, hash);
+        if (indexMetaData.isRoutingPartitionedIndex()) {
+            partitionOffset = Math.floorMod(Murmur3HashFunction.hash(id), indexMetaData.getRoutingPartitionSize());
+        } else {
+            // we would have still got 0 above but this check just saves us an unnecessary hash calculation
```

Comment:
I think either way we write it, we're always going to have to check this condition somewhere or just always perform the extra hash computation.

My preference is to pass the string to `calculateScaledShardId` since this method is called from two places (`computeTargetedShards` and `generateShardId`). The way it is now allows each of them to just pass a string without each caller duplicating the hash logic before calling it.